SU2_CFD/src/variables/CFEAVariable.cpp

```diff
@@ -93,46 +93,107 @@ void CFEAVariable::SetSolution_Vel_time_n() { Solution_Vel_time_n = Solution_Vel
 
 void CFEAVariable::SetSolution_Accel_time_n() { Solution_Accel_time_n = Solution_Accel; }
 
-void CFEAVariable::Register_femSolution_time_n() {
-  for (unsigned long iPoint = 0; iPoint < nPoint; iPoint++)
-    for (unsigned long iVar = 0; iVar < nVar; iVar++)
-      AD::RegisterInput(Solution_time_n(iPoint,iVar));
+void CFEAVariable::Register_femSolution_time_n(bool input, bool push_index) {
+  for (unsigned long iPoint = 0; iPoint < nPoint; ++iPoint) {
+    for(unsigned long iVar=0; iVar<nVar; ++iVar) {
+      if(input) {
+        if(push_index) {
+          AD::RegisterInput(Solution_time_n(iPoint,iVar));
+        }
+        else {
+          AD::RegisterInput(Solution_time_n(iPoint,iVar), false);
+          AD::SetIndex(AD_Time_n_InputIndex(iPoint,iVar), Solution_time_n(iPoint,iVar));
+        }
+      }
+      else {
+        AD::RegisterOutput(Solution_time_n(iPoint,iVar));
+        if(!push_index)
+          AD::SetIndex(AD_Time_n_OutputIndex(iPoint,iVar), Solution_time_n(iPoint,iVar));
+      }
+    }
+  }
 }
 
-void CFEAVariable::RegisterSolution_Vel(bool input) {
-  if (input) {
-    for (unsigned long iPoint = 0; iPoint < nPoint; iPoint++)
-      for (unsigned long iVar = 0; iVar < nVar; iVar++)
-        AD::RegisterInput(Solution_Vel(iPoint,iVar));
-  }
-  else {
-    for (unsigned long iPoint = 0; iPoint < nPoint; iPoint++)
-      for (unsigned long iVar = 0; iVar < nVar; iVar++)
+void CFEAVariable::RegisterSolution_Vel(bool input, bool push_index) {
+  for (unsigned long iPoint = 0; iPoint < nPoint; ++iPoint) {
+    for(unsigned long iVar=0; iVar<nVar; ++iVar) {
+      if(input) {
+        if(push_index) {
+          AD::RegisterInput(Solution_Vel(iPoint,iVar));
+        }
+        else {
+          AD::RegisterInput(Solution_Vel(iPoint,iVar), false);
+          AD::SetIndex(AD_Vel_InputIndex(iPoint,iVar), Solution_Vel(iPoint,iVar));
+        }
+      }
+      else {
         AD::RegisterOutput(Solution_Vel(iPoint,iVar));
+        if(!push_index)
+          AD::SetIndex(AD_Vel_OutputIndex(iPoint,iVar), Solution_Vel(iPoint,iVar));
+      }
+    }
```

Comment:
We could define a helper function to do this registration, it's starting to be repeated quite a few times now that we are adopting the new strategy.

Reply:
This can definitely be improved. I had to add this type of registration for all the time relevant solutions, so they are repeated many times.